scripts/framework-applications/notify-suppliers-with-signed-framework-agreement.py

```diff
@@ -0,0 +1,131 @@
+    """Get the email addresses for each user belonging to `supplier_id`"""
+    supplier_users = api_client.find_users_iter(supplier_id=supplier_id, personal_data_removed=False)
+    return [user["emailAddress"] for user in supplier_users if user["active"]]
+
+
+def get_previous_framework(framework: dict) -> dict:
+    """Find the framework iteration that comes before `framework`
+
+    Raises ValueError if the previous framework does not exist or cannot be
+    found.
+    """
+    # not convinced this is the best way to do this
+    framework_family = framework["family"]
+    frameworks_in_family = (fw for fw in api_client.find_frameworks()["frameworks"] if fw["family"] == framework_family)
+    if not frameworks_in_family:
+        raise ValueError(f"could not find any frameworks for family '{framework_family}'")
+    if len(frameworks_in_family) < 2:
+        raise ValueError(f"framework '{framework['slug']}' does not have a previous framework")
+
+    frameworks_in_family.sort(key=lambda fw: fw["frameworkLiveAtUTC"])
+    previous_framework = frameworks_in_family[-2]
+
+    assert previous_framework["slug"] != framework["slug"]
+
+    return previous_framework
+
+
+if __name__ == "__main__":
+    arguments = docopt(__doc__)
+
+    STAGE = arguments["<stage>"]
+    FRAMEWORK_SLUG = arguments["<framework>"]
+    NOTIFY_API_KEY = arguments["<notify_api_key>"]
+    NOTIFY_TEMPLATE_ID = arguments["<notify_template_id>"]
+    DRY_RUN = arguments["--dry-run"]
+
+    logger = logging_helpers.configure_logger()
+    mail_client = scripts_notify_client(NOTIFY_API_KEY, logger=logger)
+    api_client = DataAPIClient(base_url=get_api_endpoint_from_stage(STAGE), auth_token=get_auth_token("api", STAGE))
+
+    framework = api_client.get_framework(FRAMEWORK_SLUG).get("frameworks")
+    get_previous_framework = get_previous_framework(framework)
+
+
+    supplier_ids = get_supplier_ids_from_args(arguments)
+    if supplier_ids is None:
+        supplier_ids = get_supplier_ids_not_signed(api_client, FRAMEWORK_SLUG)
+
+    # Flatten list of lists
+    email_addresses = list(chain.from_iterable(get_email_addresses_for_supplier(api_client, supplier_id)
+                                               for supplier_id in supplier_ids))
+
+    prefix = "[Dry Run] " if DRY_RUN else ""
+    user_count = len(email_addresses)
+
+    for count, email in enumerate(email_addresses, start=1):
+        logger.info(
+            f"{prefix}Sending email to supplier user {count} of {user_count}: {hash_string(email)}"
+        )
+        if not DRY_RUN:
+            mail_client.send_email(
+                to_email_address=email,
+                template_name_or_id=NOTIFY_TEMPLATE_ID,
+                personalisation={
+                    "framework_name": framework["name"],
+                    "previous_framework_name": previous_framework["name"],
+                }
+            )
```

Comment:
I don't think we need to genericise it here, but this pattern of how to send emails to a bunch of users is starting to seem quite familiar.

Reply:
Yes, I've been thinking about that too